File: backend/src/main/java/pers/eastwind/billmanager/controller/ReimburseController.java
package pers.eastwind.billmanager.controller;

import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.web.bind.annotation.*;
import pers.eastwind.billmanager.model.dto.ActionsResult;
import pers.eastwind.billmanager.model.dto.PageResult;
import pers.eastwind.billmanager.model.dto.ReimburseQueryParam;
import pers.eastwind.billmanager.model.dto.ReimbursementDTO;
import pers.eastwind.billmanager.service.AttachmentService;
import pers.eastwind.billmanager.service.ReimburseService;

import java.util.List;

/**
 * 报销单控制器
 */
@Slf4j
@RestController
@RequestMapping("/api/reimburse")
public class ReimburseController {
    private final ReimburseService reimburseService;
    private final AttachmentService attachmentService;

    public ReimburseController(ReimburseService reimburseService, AttachmentService attachmentService) {
        this.reimburseService = reimburseService;
        this.attachmentService = attachmentService;
    }

    /**
     * 根据条件查找报销单
     *
     * @return 报销单列表
     */
    @PostMapping("/query")
    public PageResult<ReimbursementDTO> queryByParam(@RequestBody ReimburseQueryParam queryParam) {

        Page<ReimbursementDTO> pageResult = reimburseService.findByParam(queryParam);
        return PageResult.fromPage(pageResult);

    }

    /**
     * 根据 ID 获取报销单
     */
    @GetMapping("/{id}")
    public ReimbursementDTO getById(@PathVariable Integer id) {
        return reimburseService.findById(id);
    }

    /**
     * 创建报销单
     *
     * @param reimbursementDTO 报销单
     * @return 保存后的报销单
     */
    @PostMapping
    public ReimbursementDTO create(@RequestBody ReimbursementDTO reimbursementDTO) {
        return reimburseService.create(reimbursementDTO);
    }

    /**
     * 保存报销单
     *
     * @param reimbursementDTO 报销单
     * @return 保存后的报销单
     */
    @PutMapping
    public ReimbursementDTO save(@RequestBody ReimbursementDTO reimbursementDTO) {
        return reimburseService.update(reimbursementDTO);
    }

    /**
     * 删除报销单
     */
    @DeleteMapping
    public ActionsResult<Integer, Void> delete(@RequestBody List<Integer> ids) {
        return reimburseService.delete(ids);
    }

    /**
     * 处理报销单
     */
    @PutMapping("/process")
    public ActionsResult<Integer, Void> process(@RequestBody List<Integer> ids) {
        return reimburseService.process(ids);
    }

    /**
     * 处理完成报销单
     */
    @PutMapping("/finish")
    public ActionsResult<Integer, Void> finish(@RequestBody List<Integer> ids) {
        return reimburseService.finish(ids);
    }

    /**
     * 导出
     */
    @PostMapping("/export")
    public Resource export(@RequestBody List<Integer> ids) {
        return attachmentService.loadByPath(reimburseService.export(ids));
    }

}

File: backend-go/internal/module/reimburse/controller/reimburse_controller.go
package controller

import (
	model2 "backend-go/internal/module/reimburse/model"
	"backend-go/internal/module/reimburse/service"
	"backend-go/internal/pkg/errs"
	"backend-go/internal/pkg/result"
	"strconv"

	"github.com/gofiber/fiber/v2"
)

type ReimburseController struct {
	reimburseService *service.ReimburseService
}

func NewReimburseController(reimburseService *service.ReimburseService) *ReimburseController {
	return &ReimburseController{
		reimburseService: reimburseService,
	}
}

func (c *ReimburseController) QueryByParam(ctx *fiber.Ctx) error {
	var param model2.ReimburseQueryParam
	if err := ctx.BodyParser(&param); err != nil {
		return err
	}

	res, err := c.reimburseService.FindByParam(ctx.Context(), &param)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) GetByID(ctx *fiber.Ctx) error {
	idStr := ctx.Params("id")
	if idStr == "" {
		return errs.NewBizError("ID 为空")
	}
	id, err := strconv.Atoi(idStr)
	if err != nil {
		return err
	}
	res, err := c.reimburseService.FindByID(ctx.Context(), uint(id))
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) Create(ctx *fiber.Ctx) error {
	var dto model2.ReimbursementDTO
	if err := ctx.BodyParser(&dto); err != nil {
		return err
	}

	res, err := c.reimburseService.Create(ctx.Context(), &dto)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) Update(ctx *fiber.Ctx) error {
	var dto model2.ReimbursementDTO
	if err := ctx.BodyParser(&dto); err != nil {
		return err
	}

	res, err := c.reimburseService.Update(ctx.Context(), &dto)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) Delete(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.reimburseService.Delete(ctx.Context(), ids)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) Process(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.reimburseService.Process(ctx.Context(), ids)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) Finish(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.reimburseService.Finish(ctx.Context(), ids)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ReimburseController) Export(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.reimburseService.Export(ctx.Context(), ids)
	if err != nil {
		return err
	}
	return ctx.Download(res, "导出.zip")
}
